cloud/deployment/deployment_token.go

```diff
@@ -403,14 +389,334 @@ func TimeAgo(date time.Time) string {
+	}
+	tokenTypes := []astrocore.ListDeploymentApiTokensParamsTokenTypes{
+		"WORKSPACE",
+	}
+	token, err := GetDeploymentTokenFromInputOrUser(id, name, deployment, organizationID, &tokenTypes, client, iamClient)
+	if err != nil {
+		return err
+	}
+	roles := *token.Roles
+
+	apiTokenID := token.Id
+	var workspaceRole string
+	apiTokenDeploymentRoles := []astrocore.ApiTokenDeploymentRoleRequest{}
+
+	for i := range roles {
+		if roles[i].EntityId == deployment {
+			continue // this removes the role in question
+		}
+
+		if roles[i].EntityType == workspaceEntity {
+			workspaceRole = roles[i].Role
+		}
+
+		if roles[i].EntityType == deploymentEntity {
+			apiTokenDeploymentRoles = append(apiTokenDeploymentRoles, astrocore.ApiTokenDeploymentRoleRequest{
+				EntityId: roles[i].EntityId,
+				Role:     roles[i].Role,
+			})
+		}
+	}
+	updateWorkspaceAPITokenRoles := astrocore.UpdateWorkspaceApiTokenRolesRequest{
+		Deployment: &apiTokenDeploymentRoles,
+		Workspace:  &workspaceRole,
+	}
+	updateWorkspaceAPITokenRequest := astrocore.UpdateWorkspaceApiTokenRequest{
+		Name:        token.Name,
+		Description: token.Description,
+		Roles:       &updateWorkspaceAPITokenRoles,
+	}
+
+	resp, err := client.UpdateWorkspaceApiTokenWithResponse(httpContext.Background(), organizationID, workspace, apiTokenID, updateWorkspaceAPITokenRequest)
+	if err != nil {
+		return err
+	}
+	err = astrocore.NormalizeAPIError(resp.HTTPResponse, resp.Body)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(out, "Astro Workspace API token %s was successfully removed from the Deployment\n", token.Name)
+	return nil
+}
+
+func UpsertWorkspaceTokenDeploymentRole(id, name, role, workspace, deployment, operation string, out io.Writer, client astrocore.CoreClient, iamClient astrocoreiam.CoreClient) error {
+	ctx, err := context.GetCurrentContext()
+	if err != nil {
+		return err
+	}
+
+	organizationID := ctx.Organization
+
+	if workspace == "" {
+		workspace = ctx.Workspace
+	}
+
+	var token astrocoreiam.ApiToken
+	if operation == "create" {
+		tokenTypes := []astrocore.ListWorkspaceApiTokensParamsTokenTypes{"WORKSPACE"}
+		token, err = workspace2.GetTokenFromInputOrUser(id, name, workspace, organizationID, &tokenTypes, client, iamClient)
+		if err != nil {
+			return err
+		}
+	} else {
+		tokenTypes := []astrocore.ListDeploymentApiTokensParamsTokenTypes{"WORKSPACE"}
+		token, err = GetDeploymentTokenFromInputOrUser(id, name, deployment, organizationID, &tokenTypes, client, iamClient)
+		if err != nil {
+			return err
+		}
+	}
+
+	apiTokenID := token.Id
+	var workspaceRole string
+	apiTokenDeploymentRole := astrocore.ApiTokenDeploymentRoleRequest{
+		EntityId: deployment,
+		Role:     role,
+	}
+	apiTokenDeploymentRoles := []astrocore.ApiTokenDeploymentRoleRequest{apiTokenDeploymentRole}
+
+	for i := range *token.Roles {
+		roles := *token.Roles
+		if roles[i].EntityId == deployment {
+			if roles[i].Role == role {
+				return errWorkspaceTokenInDeployment
+			} else {
+				continue
```

Comment:
This is not needed.

Reply:
The continue is needed as line 614 would assign the old deployment role to the token instead of the role the user is assigning.